tao_compiler/mlir/disc/transforms/dot_rewriter.cc

```diff
@@ -190,6 +191,463 @@ struct TransposeFoldingConvert : public OpRewritePattern<mhlo::DotGeneralOp> {
+  // tokens
+  void categorizeTokens(
+      const TokensTy& tokens, llvm::SmallDenseSet<char>& contracting_tokens,
+      llvm::SmallDenseSet<char>& batching_tokens,
+      llvm::SmallDenseSet<char>& lhs_non_contracting_tokens,
+      llvm::SmallDenseSet<char>& rhs_non_contracting_tokens) const;
+  // Reorder the tokens for lhs/rhs/result into a legalized layout
+  void getReorderedTokens(llvm::SmallDenseSet<char> contracting_tokens,
+                          llvm::SmallDenseSet<char> batching_tokens,
+                          llvm::SmallDenseSet<char> lhs_non_contracting_tokens,
+                          llvm::SmallDenseSet<char> rhs_non_contracting_tokens,
+                          SmallVector<char> lhs_original_tokens,
+                          SmallVector<char> rhs_original_tokens,
+                          SmallVector<char> result_original_tokens,
+                          bool& isLhsBNC, bool& isRhsBNC,
+                          SmallVector<char>& lhs_reordered_tokens,
+                          SmallVector<char>& rhs_reordered_tokens,
+                          SmallVector<char>& result_reordered_tokens) const;
+  // Insert potentialy needed transpose and reshape for lhs/rhs
+  Value processOperand(PatternRewriter& rewriter, Value original_operand,
+                       Location loc, SmallVector<char> original_tokens,
+                       SmallVector<char> reordered_tokens,
+                       llvm::SmallDenseSet<char> non_contracting_tokens,
+                       llvm::SmallDenseSet<char> contracting_tokens,
+                       llvm::SmallDenseSet<char> batching_tokens,
+                       bool isBNC) const;
+  Value processResult(PatternRewriter& rewriter, Location loc, Value dot_result,
+                      Value orig_lhs, Value orig_rhs,
+                      llvm::SmallDenseSet<char> contracting_tokens,
+                      llvm::SmallDenseSet<char> batching_tokens,
+                      llvm::SmallDenseSet<char> lhs_non_contracting_tokens,
+                      llvm::SmallDenseSet<char> rhs_non_contracting_tokens,
+                      SmallVector<char> lhs_original_tokens,
+                      SmallVector<char> rhs_original_tokens,
+                      SmallVector<char> result_original_tokens,
+                      SmallVector<char> result_reordered_tokens) const;
+};
+
+void EinsumToDotGeneralPattern::categorizeTokens(
+    const TokensTy& tokens, llvm::SmallDenseSet<char>& contracting_tokens,
+    llvm::SmallDenseSet<char>& batching_tokens,
+    llvm::SmallDenseSet<char>& lhs_non_contracting_tokens,
+    llvm::SmallDenseSet<char>& rhs_non_contracting_tokens) const {
+  for (auto token : tokens) {
+    // is a contracing dim token, if both lhs/rhs have it, but result doesn't
+    // have it
+    if (token.second.count(kIsLhs) > 0 && token.second.count(kIsRhs) > 0 &&
+        token.second.count(kIsResult) == 0) {
+      contracting_tokens.insert(token.first);
```

Comment:
The `->` part is optional in torch.enisum. Is is possible that `token.second.count(kIsResult) == 0` just because the result part is omitted in the equation? In such case, `token` may be a batch token.

Reply:
Good comments, updated with deduction when no "->" in the equation string.